sql/driver/conn.go

```diff
@@ -0,0 +1,275 @@
+	// in most cases when the table already exists.
+	if gr, err := c.db.Get(nameKey); err != nil {
+		return nil, err
+	} else if gr.Exists() {
+		if p.IfNotExists {
+			return &rows{}, nil
+		}
+		return nil, fmt.Errorf("table \"%s\" already exists", p.Name.Name)
+	}
+
+	ir, err := c.db.Inc(keys.DescIDGenerator, 1)
+	if err != nil {
+		return nil, err
+	}
+	desc.ID = uint32(ir.ValueInt() - 1)
+
+	// TODO(pmattis): Be cognizant of error messages when this is ported to the
+	// server. The error currently returned below is likely going to be difficult
+	// to interpret.
+	err = c.db.Txn(func(txn *client.Txn) error {
+		descKey := keys.MakeDescMetadataKey(desc.ID)
+		b := &client.Batch{}
+		b.CPut(nameKey, descKey, nil)
+		b.Put(descKey, &desc)
+		return txn.Commit(b)
+	})
+	if err != nil {
+		// TODO(pmattis): Need to handle if-not-exists here as well.
+		return nil, err
+	}
+	return &rows{}, nil
+}
+
+func (c *conn) del(p *parser.Delete, args []driver.Value) (*rows, error) {
+	return nil, fmt.Errorf("TODO(pmattis): unimplemented: %T %s", p, p)
+}
+
+func (c *conn) insert(p *parser.Insert, args []driver.Value) (*rows, error) {
+	return nil, fmt.Errorf("TODO(pmattis): unimplemented: %T %s", p, p)
+}
+
+func (c *conn) sel(p *parser.Select, args []driver.Value) (*rows, error) {
+	return nil, fmt.Errorf("TODO(pmattis): unimplemented: %T %s", p, p)
+}
+
+func (c *conn) showDatabases(p *parser.ShowDatabases, args []driver.Value) (*rows, error) {
+	prefix := keys.MakeNameMetadataKey(structured.RootNamespaceID, "")
+	sr, err := c.db.Scan(prefix, prefix.PrefixEnd(), 0)
+	if err != nil {
+		return nil, err
+	}
+	names := make([]string, len(sr))
+	for i, row := range sr {
+		names[i] = string(bytes.TrimPrefix(row.Key, prefix))
+	}
+	return &rows{
+		columns: []string{"database"},
+		rows:    names,
+		pos:     -1,
+	}, nil
+}
+
+func (c *conn) showTables(p *parser.ShowTables, args []driver.Value) (*rows, error) {
+	if p.Name == "" {
+		if c.database == "" {
+			return nil, fmt.Errorf("no database specified")
+		}
+		p.Name = c.database
+	}
+	dbID, err := c.lookupDatabase(p.Name)
+	if err != nil {
+		return nil, err
+	}
+	prefix := keys.MakeNameMetadataKey(dbID, "")
+	sr, err := c.db.Scan(prefix, prefix.PrefixEnd(), 0)
+	if err != nil {
+		return nil, err
+	}
+	names := make([]string, len(sr))
+	for i, row := range sr {
+		names[i] = string(bytes.TrimPrefix(row.Key, prefix))
+	}
+	return &rows{
+		columns: []string{"tables"},
+		rows:    names,
+		pos:     -1,
+	}, nil
+}
+
+func (c *conn) update(p *parser.Update, args []driver.Value) (*rows, error) {
+	return nil, fmt.Errorf("TODO(pmattis): unimplemented: %T %s", p, p)
+}
+
+func (c *conn) useDatabase(p *parser.Use, args []driver.Value) (*rows, error) {
+	c.database = p.Name
+	return &rows{}, nil
+}
+
+func (c *conn) lookupDatabase(name string) (uint32, error) {
+	nameKey := keys.MakeNameMetadataKey(structured.RootNamespaceID, name)
```

Comment:
Yeah, too early for a TODO. Not sure if the caching would go here or elsewhere or even how it will work. Not even sure if we need caching of the `database name -> database id` mapping as usage of that mapping should be rare. We will need caching of `TableDescriptors`. @tamird is thinking hard about that.